Segment formations logic into different files and simplify category filters

ostrovaya opened the pull request:
<!-- General useful tooling:
    - [ScreenToGif](https://www.screentogif.com/): Free, open source screen recorder that can export to MP4. If the changes are visual, these can help you tell us exactly what the changes imply!
-->
<!-- Feel free to remove unused parts of this template. -->

## Description of the proposed changes
<!-- A clear and concise description (or visuals) of what the changes imply. -->
<!-- If it closes an issue, make sure to link the issue by using "(Closes/Fixes/Resolves) #(Issue Number)" in your pull request. -->

**Commits are granular for easier PR review. Note that many line changes are strictly movement of code.**

The `formations.lua` file is pretty large and hard to reason about since it contains logic that is necessary to setup formations but isn't necessary to be colocated with formations. There are many static declarations that are only used in that logic. So, a new file is made to hold that. Includes a few simple optimizations.

- Move all `categories` related declarations and logic out of `formations.lua` into its own file.
  - Import as necessary
- In the new file
  - Upvalue commonly used functions
  - Localize commonly used table fields
  - Simplify processing in `CategorizeUnits` that iterates through the `.Categories` of footprint sizes to generate filters by doing it on the first pass instead.

## Testing done on the proposed changes
<!-- List all relevant testing that you've done to confirm the changes work. -->

### Replay simulation

Rebase on `af859b5d21a8bf165ac20863557f572d216c7633`. Simulate replay #25004163 with `wld_RunWithTheWind`. Observe no desync.

Unrelated warnings:
```
WARNING: Duplicate definition of console command ""
WARNING:  NUM PROPS = 5182
WARNING: GetResource: Invalid name ""
WARNING: GetResource: Invalid name ""
WARNING: GetResource: Invalid name ""
WARNING: ACU kill detected. Rating for ranked games is now enforced.
WARNING: FactoryRebuildUnits failed to rebuild correctly for factory urb0101 (entity ID 3145834).
WARNING: Rebuild data:
WARNING: Progress: 0.050000
WARNING: BuildTime: 10.000001
WARNING: Health: 8.000000
WARNING: 20
WARNING: ACU kill detected. Rating for ranked games is now enforced.
WARNING: FactoryRebuildUnits failed to rebuild correctly for factory ueb0301 (entity ID 2097577).
WARNING: Rebuild data:
WARNING: Progress: 0.924063
WARNING: BuildTime: 1441.538696
WARNING: Health: 740.233337
WARNING: 90
WARNING: ACU kill detected. Rating for ranked games is now enforced.
WARNING: ACU kill detected. Rating for ranked games is now enforced.
```

### Benchmarking

Create [a benchmark for CategorizeUnits](https://github.com/ostrovaya/fa/blob/ebd6d69532f17a0f2bdbbce38fc1808dcf87f5fc/lua/benchmarks/formations.lua). Set Profiler parameters to run 100 iterations instead of 10000. Collect samples.

previous PR

![Screenshot from 2025-05-31 10-55-44](https://github.com/user-attachments/assets/db9c94c7-1dbb-430e-b57f-faae3d8d4f19)

fork

![Screenshot from 2025-06-05 10-11-12](https://github.com/user-attachments/assets/383bc24a-c679-4a29-a46c-045eb71c4732)

### Bytecode

Max stack += 3
Instruction count -= 44

Modify code for logging:

```lua
function AttackFormation(formationUnits)
    reprsl(debug.listcode(CategorizeUnits))
    local cachedResults = GetCachedResults(formationUnits, 'AttackFormation')
    if cachedResults then
        return cachedResults
    end
```

```
previous PR

INFO: { -- table: 1C0EF730 (2168 bytes)
INFO:   "(1587)    0 - GETUPVAL       1    0    0",
INFO:   "(1590)    1 - MOVE           2    1    0",
INFO:   "(1590)    2 - LOADNIL        3    5    0",
INFO:   "(1590)    3 - TFORPREP       2   34",
INFO:   "(1591)    4 - GETUPVAL       6    1    0",
INFO:   "(1591)    5 - GETTABLE       6    6    4",
INFO:   "(1592)    6 - MOVE           7    5    0",
INFO:   "(1592)    7 - LOADNIL        8   10    0",
INFO:   "(1592)    8 - TFORPREP       7    8",
INFO:   "(1593)    9 - GETTABLE      11    6    9",
INFO:   "(1594)   10 - GETGLOBAL     12    0",
INFO:   "(1594)   11 - MOVE          13   11    0",
INFO:   "(1594)   12 - CALL          12    2    4",
INFO:   "(1594)   13 - TFORPREP      12    1",
INFO:   "(1595)   14 - SETTABLE      11   14  251",
INFO:   "(1594)   15 - TFORLOOP      12    0    0",
INFO:   "(1595)   16 - JMP            0   -3",
INFO:   "(1592)   17 - TFORLOOP       7    0    1",
INFO:   "(1596)   18 - JMP            0  -10",
INFO:   "(1599)   19 - GETTABLE       7    6  252",
INFO:   "(1600)   20 - GETGLOBAL      8    0",
INFO:   "(1600)   21 - MOVE           9    7    0",
INFO:   "(1600)   22 - CALL           8    2    4",
INFO:   "(1600)   23 - TFORPREP       8    1",
INFO:   "(1601)   24 - SETTABLE       7   10  251",
INFO:   "(1600)   25 - TFORLOOP       8    0    0",
INFO:   "(1601)   26 - JMP            0   -3",
INFO:   "(1604)   27 - GETTABLE       8    6  253",
INFO:   "(1605)   28 - GETGLOBAL      9    0",
INFO:   "(1605)   29 - MOVE          10    8    0",
INFO:   "(1605)   30 - CALL           9    2    4",
INFO:   "(1605)   31 - TFORPREP       9    1",
INFO:   "(1606)   32 - SETTABLE       8   11  251",
INFO:   "(1605)   33 - TFORLOOP       9    0    0",
INFO:   "(1606)   34 - JMP            0   -3",
INFO:   "(1609)   35 - SETTABLE       6  254  255",
INFO:   "(1610)   36 - SETTABLE       6  256  255",
INFO:   "(1611)   37 - SETTABLE       6  257  251",
INFO:   "(1590)   38 - TFORLOOP       2    0    1",
INFO:   "(1611)   39 - JMP            0  -36",
INFO:   "(1615)   40 - MOVE           2    0    0",
INFO:   "(1615)   41 - LOADNIL        3    5    0",
INFO:   "(1615)   42 - TFORPREP       2  106",
INFO:   "(1616)   43 - LOADBOOL       6    0    0",
INFO:   "(1617)   44 - MOVE           7    1    0",
INFO:   "(1617)   45 - LOADNIL        8   10    0",
INFO:   "(1617)   46 - TFORPREP       7   92",
INFO:   "(1618)   47 - MOVE          11   10    0",
INFO:   "(1618)   48 - LOADNIL       12   14    0",
INFO:   "(1618)   49 - TFORPREP      11   84",
INFO:   "(1619)   50 - GETGLOBAL     15    8",
INFO:   "(1619)   51 - GETTABLE      16   10   13",
INFO:   "(1619)   52 - MOVE          17    5    0",
INFO:   "(1619)   53 - CALL          15    3    2",
INFO:   "(1619)   54 - TEST          15   15    0",
INFO:   "(1619)   55 - JMP            0   78",
INFO:   "(1620)   56 - SELF          15    5  259",
INFO:   "(1620)   57 - CALL          15    2    2",
INFO:   "(1621)   58 - GETGLOBAL     16   10",
INFO:   "(1621)   59 - GETTABLE      16   16  261",
INFO:   "(1621)   60 - GETTABLE      17   15  262",
INFO:   "(1621)   61 - GETTABLE      17   17  263",
INFO:   "(1621)   62 - GETTABLE      18   15  262",
INFO:   "(1621)   63 - GETTABLE      18   18  264",
INFO:   "(1621)   64 - CALL          16    3    2",
INFO:   "(1622)   65 - GETTABLE      17   15  265",
INFO:   "(1624)   66 - GETUPVAL      18    1    0",
INFO:   "(1624)   67 - GETTABLE      18   18    9",
INFO:   "(1624)   68 - GETTABLE      18   18   13",
INFO:   "(1624)   69 - GETTABLE      18   18   16",
INFO:   "(1624)   70 - TEST          18   18    1",
INFO:   "(1624)   71 - JMP            0    8",
INFO:   "(1625)   72 - GETUPVAL      18    1    0",
INFO:   "(1625)   73 - GETTABLE      18   18    9",
INFO:   "(1625)   74 - GETTABLE      18   18   13",
INFO:   "(1625)   75 - NEWTABLE      19    0    2",
INFO:   "(1625)   76 - SETTABLE      19  266  255",
INFO:   "(1625)   77 - NEWTABLE      20    0    0",
INFO:   "(1625)   78 - SETTABLE      19  267   20",
INFO:   "(1625)   79 - SETTABLE      18   16   19",
INFO:   "(1627)   80 - GETUPVAL      18    1    0",
INFO:   "(1627)   81 - GETTABLE      18   18    9",
INFO:   "(1627)   82 - GETTABLE      18   18   13",
INFO:   "(1627)   83 - GETTABLE      18   18   16",
INFO:   "(1627)   84 - GETUPVAL      19    1    0",
INFO:   "(1627)   85 - GETTABLE      19   19    9",
INFO:   "(1627)   86 - GETTABLE      19   19   13",
INFO:   "(1627)   87 - GETTABLE      19   19   16",
INFO:   "(1627)   88 - GETTABLE      19   19  266",
INFO:   "(1627)   89 - ADD           19   19  268",
INFO:   "(1627)   90 - SETTABLE      18  266   19",
INFO:   "(1628)   91 - GETUPVAL      18    1    0",
INFO:   "(1628)   92 - GETTABLE      18   18    9",
INFO:   "(1628)   93 - GETTABLE      18   18   13",
INFO:   "(1628)   94 - GETTABLE      18   18   16",
INFO:   "(1628)   95 - GETTABLE      18   18  267",
INFO:   "(1628)   96 - GETGLOBAL     19   19",
INFO:   "(1628)   97 - GETTABLE      19   19   17",
INFO:   "(1628)   98 - SETTABLE      18   17   19",
INFO:   "(1629)   99 - GETUPVAL      18    1    0",
INFO:   "(1629)  100 - GETTABLE      18   18    9",
INFO:   "(1629)  101 - GETTABLE      18   18  252",
INFO:   "(1629)  102 - GETUPVAL      19    1    0",
INFO:   "(1629)  103 - GETTABLE      19   19    9",
INFO:   "(1629)  104 - GETTABLE      19   19  252",
INFO:   "(1629)  105 - GETTABLE      19   19   16",
INFO:   "(1629)  106 - TEST          19   19    1",
INFO:   "(1629)  107 - JMP            0    1",
INFO:   "(1629)  108 - LOADK         19    5",
INFO:   "(1629)  109 - ADD           19   19  268",
INFO:   "(1629)  110 - SETTABLE      18   16   19",
INFO:   "(1631)  111 - EQ             0   13  270",
INFO:   "(1631)  112 - JMP            0   12",
INFO:   "(1632)  113 - GETGLOBAL     18   21",
INFO:   "(1632)  114 - LOADK         19   22",
INFO:   "(1632)  115 - GETGLOBAL     20   23",
INFO:   "(1632)  116 - SELF          21    5  259",
INFO:   "(1632)  117 - CALL          21    2    2",
INFO:   "(1632)  118 - GETTABLE      21   21  265",
INFO:   "(1632)  119 - CALL          20    2    2",
INFO:   "(1632)  120 - LOADK         21   24",
INFO:   "(1632)  121 - MOVE          22    9    0",
INFO:   "(1632)  122 - LOADK         23   25",
INFO:   "(1632)  123 - CONCAT        19   19   23",
INFO:   "(1632)  124 - CALL          18    2    1",
INFO:   "(1634)  125 - GETUPVAL      18    1    0",
INFO:   "(1634)  126 - GETTABLE      18   18    9",
INFO:   "(1634)  127 - GETUPVAL      19    1    0",
INFO:   "(1634)  128 - GETTABLE      19   19    9",
INFO:   "(1634)  129 - GETTABLE      19   19  254",
INFO:   "(1634)  130 - ADD           19   19  268",
INFO:   "(1634)  131 - SETTABLE      18  254   19",
INFO:   "(1635)  132 - LOADBOOL       6    1    0",
INFO:   "(1636)  133 - JMP            0    2",
INFO:   "(1618)  134 - TFORLOOP      11    0    1",
INFO:   "(1637)  135 - JMP            0  -86",
INFO:   "(1640)  136 - TEST           6    6    0",
INFO:   "(1640)  137 - JMP            0    1",
INFO:   "(1641)  138 - JMP            0    2",
INFO:   "(1617)  139 - TFORLOOP       7    0    1",
INFO:   "(1642)  140 - JMP            0  -94",
INFO:   "(1644)  141 - TEST           6    6    1",
INFO:   "(1644)  142 - JMP            0    6",
INFO:   "(1645)  143 - GETGLOBAL      7   26",
INFO:   "(1645)  144 - LOADK          8   22",
INFO:   "(1645)  145 - GETTABLE       9    5  277",
INFO:   "(1645)  146 - LOADK         10   28",
INFO:   "(1645)  147 - CONCAT         8    8   10",
INFO:   "(1645)  148 - CALL           7    2    1",
INFO:   "(1615)  149 - TFORLOOP       2    0    1",
INFO:   "(1646)  150 - JMP            0 -108",
INFO:   "(1650)  151 - MOVE           2    1    0",
INFO:   "(1650)  152 - LOADNIL        3    5    0",
INFO:   "(1650)  153 - TFORPREP       2   36",
INFO:   "(1651)  154 - MOVE           6    5    0",
INFO:   "(1651)  155 - LOADNIL        7    9    0",
INFO:   "(1651)  156 - TFORPREP       6   31",
INFO:   "(1652)  157 - GETUPVAL      10    1    0",
INFO:   "(1652)  158 - GETTABLE      10   10    4",
INFO:   "(1652)  159 - GETTABLE      10   10    8",
INFO:   "(1652)  160 - TEST          10   10    0",
INFO:   "(1652)  161 - JMP            0   26",
INFO:   "(1653)  162 - GETUPVAL      10    1    0",
INFO:   "(1653)  163 - GETTABLE      10   10    4",
INFO:   "(1653)  164 - GETTABLE      10   10    8",
INFO:   "(1653)  165 - LOADNIL       11   13    0",
INFO:   "(1653)  166 - TFORPREP      10   19",
INFO:   "(1654)  167 - LOADNIL       14   14    0",
INFO:   "(1655)  168 - GETTABLE      15   13  267",
INFO:   "(1655)  169 - LOADNIL       16   18    0",
INFO:   "(1655)  170 - TFORPREP      15    5",
INFO:   "(1656)  171 - TEST          14   14    1",
INFO:   "(1656)  172 - JMP            0    2",
INFO:   "(1657)  173 - MOVE          14   18    0",
INFO:   "(1657)  174 - JMP            0    1",
INFO:   "(1659)  175 - ADD           14   14   18",
INFO:   "(1655)  176 - TFORLOOP      15    0    1",
INFO:   "(1660)  177 - JMP            0   -7",
INFO:   "(1662)  178 - GETUPVAL      15    1    0",
INFO:   "(1662)  179 - GETTABLE      15   15    4",
INFO:   "(1662)  180 - GETTABLE      15   15    8",
INFO:   "(1662)  181 - NEWTABLE      16    0    2",
INFO:   "(1662)  182 - GETTABLE      17   13  266",
INFO:   "(1662)  183 - SETTABLE      16  266   17",
INFO:   "(1662)  184 - SETTABLE      16  279   14",
INFO:   "(1662)  185 - SETTABLE      15   12   16",
INFO:   "(1653)  186 - TFORLOOP      10    0    1",
INFO:   "(1662)  187 - JMP            0  -21",
INFO:   "(1651)  188 - TFORLOOP       6    0    1",
INFO:   "(1664)  189 - JMP            0  -33",
INFO:   "(1650)  190 - TFORLOOP       2    0    1",
INFO:   "(1665)  191 - JMP            0  -38",
INFO:   "(1668)  192 - GETGLOBAL      2   30",
INFO:   "(1668)  193 - GETUPVAL       3    1    0",
INFO:   "(1668)  194 - CALL           2    2    1",
INFO:   "(1670)  195 - GETUPVAL       2    1    0",
INFO:   "(1670)  196 - RETURN         2    2    0",
INFO:   "(1671)  197 - RETURN         0    1    0",
INFO:   maxstack = 24,
INFO:   numparams = 1
INFO: }
```

```
fork

INFO: { -- table: 17A03B68 (2168 bytes)
INFO:   "( 235)    0 - GETUPVAL       1    0    0",
INFO:   "( 238)    1 - GETGLOBAL      2    0",
INFO:   "( 238)    2 - MOVE           3    1    0",
INFO:   "( 238)    3 - CALL           2    2    5",
INFO:   "( 238)    4 - TFORPREP       2   35",
INFO:   "( 239)    5 - GETUPVAL       6    1    0",
INFO:   "( 239)    6 - GETTABLE       6    6    4",
INFO:   "( 240)    7 - GETGLOBAL      7    0",
INFO:   "( 240)    8 - MOVE           8    5    0",
INFO:   "( 240)    9 - CALL           7    2    5",
INFO:   "( 240)   10 - TFORPREP       7    8",
INFO:   "( 241)   11 - GETTABLE      11    6    9",
INFO:   "( 242)   12 - GETGLOBAL     12    0",
INFO:   "( 242)   13 - MOVE          13   11    0",
INFO:   "( 242)   14 - CALL          12    2    4",
INFO:   "( 242)   15 - TFORPREP      12    1",
INFO:   "( 243)   16 - SETTABLE      11   14  251",
INFO:   "( 242)   17 - TFORLOOP      12    0    0",
INFO:   "( 243)   18 - JMP            0   -3",
INFO:   "( 240)   19 - TFORLOOP       7    0    1",
INFO:   "( 244)   20 - JMP            0  -10",
INFO:   "( 247)   21 - GETTABLE       7    6  252",
INFO:   "( 248)   22 - GETGLOBAL      8    0",
INFO:   "( 248)   23 - MOVE           9    7    0",
INFO:   "( 248)   24 - CALL           8    2    4",
INFO:   "( 248)   25 - TFORPREP       8    1",
INFO:   "( 249)   26 - SETTABLE       7   10  251",
INFO:   "( 248)   27 - TFORLOOP       8    0    0",
INFO:   "( 249)   28 - JMP            0   -3",
INFO:   "( 252)   29 - GETTABLE       8    6  253",
INFO:   "( 253)   30 - GETGLOBAL      9    0",
INFO:   "( 253)   31 - MOVE          10    8    0",
INFO:   "( 253)   32 - CALL           9    2    4",
INFO:   "( 253)   33 - TFORPREP       9    1",
INFO:   "( 254)   34 - SETTABLE       8   11  251",
INFO:   "( 253)   35 - TFORLOOP       9    0    0",
INFO:   "( 254)   36 - JMP            0   -3",
INFO:   "( 257)   37 - SETTABLE       6  254  255",
INFO:   "( 258)   38 - SETTABLE       6  256  255",
INFO:   "( 259)   39 - SETTABLE       6  257  251",
INFO:   "( 238)   40 - TFORLOOP       2    0    1",
INFO:   "( 259)   41 - JMP            0  -37",
INFO:   "( 263)   42 - GETGLOBAL      2    0",
INFO:   "( 263)   43 - MOVE           3    0    0",
INFO:   "( 263)   44 - CALL           2    2    5",
INFO:   "( 263)   45 - TFORPREP       2  100",
INFO:   "( 265)   46 - LOADBOOL       6    0    0",
INFO:   "( 267)   47 - GETGLOBAL      7    0",
INFO:   "( 267)   48 - MOVE           8    1    0",
INFO:   "( 267)   49 - CALL           7    2    5",
INFO:   "( 267)   50 - TFORPREP       7   85",
INFO:   "( 268)   51 - GETUPVAL      11    1    0",
INFO:   "( 268)   52 - GETTABLE      11   11    9",
INFO:   "( 270)   53 - GETGLOBAL     12    0",
INFO:   "( 270)   54 - MOVE          13   10    0",
INFO:   "( 270)   55 - CALL          12    2    5",
INFO:   "( 270)   56 - TFORPREP      12   74",
INFO:   "( 272)   57 - GETGLOBAL     16    8",
INFO:   "( 272)   58 - GETTABLE      17   10   14",
INFO:   "( 272)   59 - MOVE          18    5    0",
INFO:   "( 272)   60 - CALL          16    3    2",
INFO:   "( 272)   61 - TEST          16   16    0",
INFO:   "( 272)   62 - JMP            0   68",
INFO:   "( 273)   63 - SELF          16    5  259",
INFO:   "( 273)   64 - CALL          16    2    2",
INFO:   "( 274)   65 - GETUPVAL      17    2    0",
INFO:   "( 274)   66 - GETTABLE      18   16  260",
INFO:   "( 274)   67 - GETTABLE      18   18  261",
INFO:   "( 274)   68 - GETTABLE      19   16  260",
INFO:   "( 274)   69 - GETTABLE      19   19  262",
INFO:   "( 274)   70 - CALL          17    3    2",
INFO:   "( 276)   71 - TEST          17   17    1",
INFO:   "( 276)   72 - JMP            0   11",
INFO:   "( 277)   73 - GETGLOBAL     18   13",
INFO:   "( 277)   74 - LOADK         19   14",
INFO:   "( 277)   75 - GETGLOBAL     20   15",
INFO:   "( 277)   76 - SELF          21    5  259",
INFO:   "( 277)   77 - CALL          21    2    2",
INFO:   "( 277)   78 - GETTABLE      21   21  266",
INFO:   "( 277)   79 - CALL          20    2    2",
INFO:   "( 277)   80 - LOADK         21   17",
INFO:   "( 277)   81 - CONCAT        19   19   21",
INFO:   "( 277)   82 - CALL          18    2    1",
INFO:   "( 278)   83 - LOADK         17    5",
INFO:   "( 281)   84 - GETTABLE      18   16  266",
INFO:   "( 284)   85 - GETTABLE      19   11   14",
INFO:   "( 286)   86 - GETTABLE      20   19   17",
INFO:   "( 286)   87 - TEST          20   20    1",
INFO:   "( 286)   88 - JMP            0    6",
INFO:   "( 287)   89 - NEWTABLE      20    0    2",
INFO:   "( 287)   90 - SETTABLE      20  268  255",
INFO:   "( 287)   91 - GETGLOBAL     21   20",
INFO:   "( 287)   92 - GETTABLE      21   21   18",
INFO:   "( 287)   93 - SETTABLE      20  269   21",
INFO:   "( 287)   94 - SETTABLE      19   17   20",
INFO:   "( 290)   95 - GETTABLE      20   19   17",
INFO:   "( 292)   96 - GETTABLE      21   20  268",
INFO:   "( 292)   97 - ADD           21   21  271",
INFO:   "( 292)   98 - SETTABLE      20  268   21",
INFO:   "( 293)   99 - GETTABLE      21   20  269",
INFO:   "( 293)  100 - GETGLOBAL     22   20",
INFO:   "( 293)  101 - GETTABLE      22   22   18",
INFO:   "( 293)  102 - ADD           21   21   22",
INFO:   "( 293)  103 - SETTABLE      20  269   21",
INFO:   "( 294)  104 - GETTABLE      21   11  252",
INFO:   "( 294)  105 - GETTABLE      22   11  252",
INFO:   "( 294)  106 - GETTABLE      22   22   17",
INFO:   "( 294)  107 - TEST          22   22    1",
INFO:   "( 294)  108 - JMP            0    1",
INFO:   "( 294)  109 - LOADK         22    5",
INFO:   "( 294)  110 - ADD           22   22  271",
INFO:   "( 294)  111 - SETTABLE      21   17   22",
INFO:   "( 296)  112 - EQ             0   14  272",
INFO:   "( 296)  113 - JMP            0   12",
INFO:   "( 297)  114 - GETGLOBAL     21   23",
INFO:   "( 297)  115 - LOADK         22   14",
INFO:   "( 297)  116 - GETGLOBAL     23   15",
INFO:   "( 297)  117 - SELF          24    5  259",
INFO:   "( 297)  118 - CALL          24    2    2",
INFO:   "( 297)  119 - GETTABLE      24   24  266",
INFO:   "( 297)  120 - CALL          23    2    2",
INFO:   "( 297)  121 - LOADK         24   24",
INFO:   "( 297)  122 - MOVE          25    9    0",
INFO:   "( 297)  123 - LOADK         26   25",
INFO:   "( 297)  124 - CONCAT        22   22   26",
INFO:   "( 297)  125 - CALL          21    2    1",
INFO:   "( 299)  126 - GETTABLE      21   11  254",
INFO:   "( 299)  127 - ADD           21   21  271",
INFO:   "( 299)  128 - SETTABLE      11  254   21",
INFO:   "( 300)  129 - LOADBOOL       6    1    0",
INFO:   "( 301)  130 - JMP            0    2",
INFO:   "( 270)  131 - TFORLOOP      12    0    1",
INFO:   "( 302)  132 - JMP            0  -76",
INFO:   "( 305)  133 - TEST           6    6    0",
INFO:   "( 305)  134 - JMP            0    1",
INFO:   "( 306)  135 - JMP            0    2",
INFO:   "( 267)  136 - TFORLOOP       7    0    1",
INFO:   "( 307)  137 - JMP            0  -87",
INFO:   "( 309)  138 - TEST           6    6    1",
INFO:   "( 309)  139 - JMP            0    6",
INFO:   "( 310)  140 - GETGLOBAL      7   13",
INFO:   "( 310)  141 - LOADK          8   14",
INFO:   "( 310)  142 - GETTABLE       9    5  276",
INFO:   "( 310)  143 - LOADK         10   27",
INFO:   "( 310)  144 - CONCAT         8    8   10",
INFO:   "( 310)  145 - CALL           7    2    1",
INFO:   "( 263)  146 - TFORLOOP       2    0    1",
INFO:   "( 311)  147 - JMP            0 -102",
INFO:   "( 314)  148 - GETGLOBAL      2   28",
INFO:   "( 314)  149 - GETUPVAL       3    1    0",
INFO:   "( 314)  150 - CALL           2    2    1",
INFO:   "( 316)  151 - GETUPVAL       2    1    0",
INFO:   "( 316)  152 - RETURN         2    2    0",
INFO:   "( 317)  153 - RETURN         0    1    0",
INFO:   maxstack = 27,
INFO:   numparams = 1
INFO: }
```

## Additional context
<!-- Add any other context about the pull request here. -->


## Checklist
- [x] Changes are annotated, including comments where useful
- [x] Changes are documented in a changelog snippet according to the [guidelines](https://faforever.github.io/fa/development/changelog#format-of-a-snippet).
- [x] Request 2-3 reviewers from the [list of reviewers and their areas of knowledge](https://github.com/FAForever/fa/blob/deploy/fafdevelop/CONTRIBUTING.md#reviewers).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Reorganized formation logic into shared modules for clearer structure
  * Simplified unit category filtering and improved formation categorization for Land, Air, Naval, and Sub types
  * Updated guard formation handling for shield vs non-shield units

* **Documentation**
  * Added changelog entry documenting these changes
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
create formationsCategories file

Commit:
migrate LandCategories

Commit:
migrate AirCategories

Commit:
migrate NavalCategories and SubCategories

Commit:
migrate CategorizeUnits and CalculateSizes

Commit:
Remove unnecessary params

Commit:
localize static table

Commit:
remove unnecessary calculations for category filters

Commit:
add missing import & comments

Commit:
upvalue math functions

Commit:
localize footprintSizeData

Commit:
localize typeData

Commit:
rename ShieldCat to ShieldCategory

Commit:
Precompute NonShieldCategory

Commit:
better spacing

Commit:
fix incorrect naming

Commit:
move TypeGroups closer to relevant context

Commit:
localize categoryData

Commit:
localize typeData

Commit:
snippet

ostrovaya commented:
Sorry for the repeat mention - I used the wrong branch on my fork.

@lL1l1 @Garanas for review.

Commit:
fix snippet

Garanas reviewed (commented):
We have the shared folder to indicate that the Lua code is used in both the UI and the simulation. See also the last sentence in my comment here:

- https://github.com/FAForever/fa/pull/6826#pullrequestreview-2885307652

The formation code is used by both Lua environments. It makes sense to put it there I think.

Garanas commented on lua/formationsCategories.lua:
```diff
@@ -0,0 +1,299 @@
+
+    AEngineer = EngineerAir,
+
+    RemainingCategory = categories.AIR - (GroundAttackAir + TransportationAir + BomberAir + AAAir + AntiNavyAir + IntelAir + ExperimentalAir + EngineerAir)
```
Unrelated to your pull request - but is it even possible for a unit to fit this category? All the other categories (`GroundAttackAir`, etc) have `categories.AIR` as part of its definition. If we then subtract that from `categories.AIR` we end up with....  🤔 !

lL1l1 replied:
`AIR - (AIR * BOMBER)` is basically `AIR - BOMBER`. It doesn't exclude all air units, so this category includes any air units that don't have typical air unit capabilities (ex: shield bubble, stealth field, antinuke, artillery).

Garanas reviewed (dismissed):
I don't like https://github.com/FAForever/fa/pull/6847/commits/0fdb4161f2edc16d83d5830553a5b47852b30413.

It introduces hidden state that makes it less trivial to test and understand the code. Ideally all functions are pure: their only input are the parameters. No global state is allowed. And their only 'effect' is their output.

In practice, we can't have code that is this pure everywhere. As an example, the majority of classes have hidden state (through the `self` value). 

But in this case we can have it be pure. And I think it should rely on just the parameter here, instead of hidden state.

ostrovaya commented:
> We have the shared folder to indicate that the Lua code is used in both the UI and the simulation. See also the last sentence in my comment here:
> 
>     * [[performance][formations] Reuse unit categorization table instead of allocating a new one when issuing uncached formation orders #6826 (review)](https://github.com/FAForever/fa/pull/6826#pullrequestreview-2885307652)
> 
> 
> The formation code is used by both Lua environments. It makes sense to put it there I think.

Sounds good! I missed that when I made a set of goals for my notes - my apologies.

> I don't like [0fdb416](https://github.com/FAForever/fa/commit/0fdb4161f2edc16d83d5830553a5b47852b30413).
> 
> It introduces hidden state that makes it less trivial to test and understand the code. Ideally all functions are pure: their only input are the parameters. No global state is allowed. And their only 'effect' is their output.
> 
> In practice, we can't have code that is this pure everywhere. As an example, the majority of classes have hidden state (through the `self` value).
> 
> But in this case we can have it be pure. And I think it should rely on just the parameter here, instead of hidden state.

Also good. Yes, let's make this a pure function instead of relying on side effects.

I'll give a shout when the changes are done :)

Commit:
make CalculateSizes more pure

Commit:
add missing annotations

Commit:
move formationsCategories to shared/formations/categorizeUnits

Commit:
add annotations for TypeGroups

Commit:
add some basic missing annotations

Commit:
further specify type annotation

Commit:
fix comment ordering

Commit:
use prior function invocation

Commit:
use pairs to support better annotations

Commit:
Add footprint size nil check, message, and override

Commit:
rename variable for clarity

ostrovaya commented:
@Garanas thanks for the feedback. I have completed the changes requested and added some annotations.

I messed up my local branches and had to force push but **relevant changes are on commit `2a43bc869237f14f201254ae407a891350a1b10b` and after.**

On annotations: 

- **I changed for iterating with `pairs` throughout**  to save 1-2 lines of `--@param` declaration for each iterator line and enable better annotations by default. This let me catch the `fs == nil` check in a following commit. Keep in mind that benchmarks show this approach is ~0.54% slower (https://github.com/faforever/fa/blob/653be9b4d35f0cf6ed9f513201d18de05513e17e/lua/benchmarks/table-loops.lua)

- I tried annotating `UnitsList` but the definition will ultimately balloon into some massive combination of duplicating each of the keys in the `...Categories` to be used for an alias to be used in some other keys which are combined with the extra fields `FootprintCounts`, `FootprintSizes`, etc. so we end up with some really gross definition that - even if the keys were specific to each layer - would still end up as something like 

`table<[LandCategoryFields | ExtraFields], [number | table | {Count: number, Filter: EntityCategory}]> `

which is somewhat gnarly. I don't see high value here for needing to add >100 lines of annotations. There will be more value if some of the code can be cleaned up first. I can add them in - no problem - but it might be better to wait on that for now.

Commit:
better format nil check

ostrovaya commented:
Actually! I got a reasonably set of annotations to make this work - just had to get a hang of the syntax of it all. Will update with the same.

Commit:
add formation data annotations

Commit:
Replace comment headings with regions and add comments

ostrovaya commented:
Will be sorting out my local git workflow to prevent these force pushes after whatever open PRs are merged/rejected but none of the commits themselves have been changed.

@Garanas I have updated the branch with your suggestions and included some annotations to help make sense of the code that much more. I ran through the test plan again with expected results and updated the byte code to match the latest commit.

ostrovaya commented:
Hmmm maybe I should also break this up into a few smaller PRs like the file changes, the annotations, and the algorithm changes. I'll do that in a few days if this isn't merged by then :)

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
```diff
@@ -0,0 +1,401 @@
+                    local footprintSizeData = categoryData[fs]
+
+                    footprintSizeData.Count = footprintSizeData.Count + 1
+                    footprintSizeData.Filter = footprintSizeData.Filter + categories[id]
```
Isn't this double adding `categories[id]` the first time `categoryData[fs]` is created?
I'm reviewing from the first commit https://github.com/FAForever/fa/pull/6847/commits/4394600b344a5fbf535770191b4ed915a3b4ca4f where the original was deleted and it had:

https://github.com/FAForever/fa/blob/8a1e045290d2003eb72de7a993f7b9f2a2385249/lua/formationsCategories.lua#L238-L246

ostrovaya replied:
That's correct. Since re-adding the category doesn't change the `Filter` and placing the addition codeblock into an `else` would increase bytecode, this seemed to be fine. I'll add a comment to be explicit about the duplication.

lL1l1 replied:
I don't think the bytecode size is an issue. It's probably worse to do an unnecessary category union operation.

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
- This can't be in /shared/ because it uses the Sim Unit specific field `UnitId`. That's just one easy fix to `Unit:GetUnitId()`.
- The annotations should also be `Unit[] | UserUnit[]` if it can actually be shared.
  - formations.lua should also be that way if the UI ever ends up using it for some new feature like a formation preview when you click to give an order.

Garanas replied:
You're right - it is probably part of a refactor in the past.

ostrovaya replied:
I updated the annotations and calls as suggested

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
```diff
@@ -0,0 +1,401 @@
+---@type EntityCategory # explicitly EntityCategory to fix unknown annotation
+NonShieldCategory = categories.ALLUNITS - ShieldCategory
+
+---@alias LandCategoryNames
+---| "Shield"
+---| "Bot1"
+---| "Bot2"
+---| "Bot3"
+---| "Bot4"
+---| "Tank1"
+---| "Tank2"
+---| "Tank3"
+---| "Tank4"
+---| "Sniper1"
+---| "Sniper2"
+---| "Sniper3"
+---| "Sniper4"
+---| "Art1"
+---| "Art2"
+---| "Art3"
+---| "Art4"
+---| "AA1"
+---| "AA2"
+---| "AA3"
+---| "Com1"
+---| "Com2"
+---| "Com3"
+---| "Com4"
+---| "Util1"
+---| "Util2"
+---| "Util3"
+---| "Util4"
+---| "RemainingCategory"
+---@type table<LandCategoryNames, EntityCategory>
+LandCategories = {
```
[`---@enum`](https://luals.github.io/wiki/annotations/#enum) exists for this use case but we don't have it working it seems.

ostrovaya replied:
Yeah, I saw that in the latest version of the plugin but noted that it doesn't exist in this repo

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
```diff
@@ -0,0 +1,401 @@
+        end
+    end
+
+    CalculateSizes(UnitsList)
```
``Undefined global `CalculateSizes`.Lua Diagnostics.(undefined-global)``
The function should be below `CalculateSizes` or else it breaks the ability to use F12 (Go to Definition) to jump around, which is quite nice to have in my opinion.

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
```diff
@@ -0,0 +1,401 @@
+end
+
+---@alias FormationLayers "Land" | "Air" | "Naval" | "Subs"
+---@alias TypeGroupData { GridSizeFraction: number, GridSizeAbsolute: integer, MinSeparationFraction: number, Types: table<integer, FormationLayers> }
```
```suggestion
---@alias TypeGroupData { GridSizeFraction: number, GridSizeAbsolute: integer, MinSeparationFraction: number, Types: FormationLayers[] }
```

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
```diff
@@ -0,0 +1,401 @@
+--#region Table initialization
+---@alias FormationLayerFootprints table<number, { Count: integer, Filter: EntityCategory }>
+---@alias FormationLayerFootprintTable { Count: integer, Filter: EntityCategory }
+---@alias FormationLayerCommonData { FootprintSizes: table<number, integer>, FootprintCounts: table<number, integer>, UnitTotal: integer, AreaTotal: number, Scale: number? }
+
+-- Table for formation categorization
+---@class FormationData
+---@field Land table<LandCategoryNames, FormationLayerFootprints> | FormationLayerCommonData
+---@field Air table<AirCategoryNames, FormationLayerFootprints> | FormationLayerCommonData
+---@field Naval table<NavalCategoryNames, FormationLayerFootprints> | FormationLayerCommonData
+---@field Subs table<SubCategoryNames, FormationLayerFootprints> | FormationLayerCommonData
```
I was trying out making a generic class (like `table<K,V> = {[K], V}`) so that there could be a better semblance of inheritance but it didn't work. Don't think there's any way to improve this then since making 4 classes for each layer type seems a bit excessive.

Garanas replied:
This is fine - the underlying data is not generic. It's very specific. This solution is fine I think.

ostrovaya replied:
I tried a generic approach also but anything that came close to modeling it that way also became pretty unwieldy to read if it worked to begin with. I'll keep it this way for now.

lL1l1 commented on lua/shared/formations/categorizeUnits.lua:
```diff
@@ -0,0 +1,401 @@
+---@type EntityCategory # explicitly EntityCategory to fix unknown annotation
+ShieldCategory = categories.uel0307 + categories.ual0307 + categories.xsl0307
+---@type EntityCategory # explicitly EntityCategory to fix unknown annotation
+NonShieldCategory = categories.ALLUNITS - ShieldCategory
```
Instead of defining the type here you can define the categories table https://github.com/FAForever/fa/blob/135e4a089367529d30ab0e4bb45011fb7a3955a8/engine/Core/Categories.lua#L7
as `---@type table<CategoryName | UnitId, EntityCategory>`.

lL1l1 reviewed (changes requested):
I looked through code changes commit by commit (helps a lot) and looked at the annotations on my local copy in vscode.

Garanas committed:
Update lua/shared/formations/categorizeUnits.lua

Co-authored-by: lL1l1 <82986251+lL1l1@users.noreply.github.com>

Garanas reviewed (commented):
With a few small tweaks as suggested by @lL1l1 we can merge this in.

Commit:
annotate categories table

Commit:
reorder CalculateSizes for annotation support

Commit:
use GetUnitId() instead of .UnitId

Commit:
add UserUnit[] to annotations

Commit:
use local unit.bp

Commit:
clarify duplication of category filter

ostrovaya commented:
I amended the PR as suggested or otherwise addressed the suggestions. I appreciate the feedback here.

@Garanas Let me know if it's satisfactory or requires further revision. Thanks

lL1l1 committed:
Merge branch 'develop' into pr/6847

lL1l1 committed:
Remove unnecessary category union op

lL1l1 committed:
Remove `pairs()` calls

improves performance and formation performance is important

FAForeverBot committed:
Update lobby changelogs

lL1l1 committed:
Fix undefined globals mistake from merge

lL1l1 committed:
Merge branch 'formations/files' of https://github.com/ostrovaya/fa into pr/6847

lL1l1 committed:
Fix annotation of Shield land category name

lL1l1 committed:
Remove footprint size nil check

it was added because intellisense warns about Footprint being nil, but the engine always gives units a footprint based on their hitbox size after init
